File: ruby/src/target-sum.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def find_target_sum_ways(nums, target)
  cache = Array.new(nums.size) { {} }
  recur(nums, cache, target, 0)
end

def recur(nums, cache, target, i)
  return 1 if i >= nums.size && target == 0
  return 0 if i >= nums.size

  cache[i][target] ||= begin
    recur(nums, cache, target - nums[i], i + 1) +
      recur(nums, cache, target + nums[i], i + 1)
  end
end

File: python3/target-sum.py
from functools import cache

class Solution:
    """
    >>> Solution().findTargetSumWays([1,1,1,1,1], 3)
    5
    >>> Solution().findTargetSumWays([1], 1)
    1
    """
    def findTargetSumWays(self, nums: list[int], target: int) -> int:
        @cache
        def recur(i, expr_result) -> int:
            if i >= len(nums):
                if expr_result == target:
                    return 1
                else:
                    return 0
            return recur(i + 1, expr_result + nums[i]) + recur(i + 1, expr_result - nums[i])
        return recur(0, 0)

